syz-fuzzer/machine_info.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sort"
+	"strings"
+)
+
+func CollectMachineInfo() ([]byte, error) {
+	if runtime.GOOS != "linux" {
+		return nil, nil
+	}
+
+	type machineInfoFunc struct {
+		name string
+		fn   func() ([]byte, error)
+	}
+
+	allMachineInfo := []machineInfoFunc{
+		{"CPU Info", readCPUInfo},
+		{"KVM", readKVMInfo},
+	}
+
+	buffer := new(bytes.Buffer)
+
+	for _, pair := range allMachineInfo {
+		fmt.Fprintf(buffer, "[%s]\n", pair.name)
+		data, err := pair.fn()
+		if err != nil {
+			if os.IsNotExist(err) {
+				buffer.WriteString(err.Error() + "\n")
+			}
+			return nil, err
+		}
+		buffer.Write(data)
+		fmt.Fprintf(buffer, "-----------------------------------\n\n")
+	}
+
+	return buffer.Bytes(), nil
+}
+
+func readCPUInfo() ([]byte, error) {
+	file, err := os.Open("/proc/cpuinfo")
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	result := scanCPUInfo(scanner)
+	return result, nil
+}
+
+func scanCPUInfo(scanner *bufio.Scanner) []byte {
+	keyOrder := make(map[string]int)
+	info := make(map[string][]string)
+	for scanner.Scan() {
+		splitted := strings.Split(scanner.Text(), ":")
+		if len(splitted) != 2 {
+			continue
+		}
+		key := strings.TrimSpace(splitted[0])
+		val := strings.TrimSpace(splitted[1])
+		if _, ok := keyOrder[key]; !ok {
+			keyIdx := len(keyOrder)
+			keyOrder[key] = keyIdx
+		}
+		info[key] = append(info[key], val)
+	}
+	keys := make([]string, 0, len(keyOrder))
+	for key := range keyOrder {
+		keys = append(keys, key)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keyOrder[keys[i]] < keyOrder[keys[j]]
+	})
+
+	buffer := new(bytes.Buffer)
+	for _, key := range keys {
+		// It is guaranteed that len(vals) >= 1
+		vals := info[key]
+		if allEqual(vals) {
+			fmt.Fprintf(buffer, "%-20s: %s\n", key, vals[0])
+		} else {
+			fmt.Fprintf(buffer, "%-20s: %s\n", key, strings.Join(vals, ", "))
+		}
+	}
```

Comment:
There are duplicated keys in the machine information and I want to group them together, so I used a hashmap to check if the key appeared previously. I agree that maybe it is not worth using such complex handling because there are probably only a hundred lines in the cpu info output.

No matter what, the sorting is definitely not necessary and that is an overlook. I think I could keep `keyOrder` and make `info` a list, and we can remove the sorting part.

Reply:
So I kept `keyOrder` but remove the sorting part. Another way is to remove `keyOrder` and find the index of the key in the array `info`. Because there are not so many keys (probably less than 30) so I don't think it will hurt the performance. Yet, go doesn't have a find or indexOf function, it also does not improve readability much or simplify the code, so I will just use an hash map.